### ISSUE TITLE
[Summary] Add gradient-based attribution methods

### ISSUE THREAD

**Issue by gsarti** [inseq-team/inseq#122]
## 🚀 Feature Request

The following is a non-exhaustive list of gradient-based feature attribution methods that could be added to the library:

<table>
<tr>
	<td> <b>Method name</b> </td>
	<td> <b>Source</b> </td>
	<td> <b>In Captum</b> </td>
	<td> <b>Code implementation</b> </td>
	<td> <b>Status</b></td>
<tr>
	<td> <a href="https://captum.ai/api/deep_lift_shap.html"><b>DeepLiftSHAP</b></a></td>
	<td> - </td>
	<td> ✅ </td>
	<td> <a href="https://github.com/pytorch/captum"><code>pytorch/captum</code></a> </td>
	<td>  </td>
<tr>
	<td> <a href="https://captum.ai/api/gradient_shap.html"><b>GradientSHAP</b></a><sup><a href="#fnote1" id="fnote1-ref" data-footnote-ref="" aria-describedby="footnote-label">1</a></sup></td>
	<td> <a href="https://papers.nips.cc/paper/7062-a-unified-approach-to-interpreting-model-predictions">Lundberg and Lee '17</a> </td>
	<td> ✅ </td>
	<td> <a href="https://github.com/pytorch/captum"><code>pytorch/captum</code></a> </td>
	<td>  </td>
<tr>
	<td> <a href="https://captum.ai/api/guided_backprop.html"><b> Guided Backprop </b></a> </td>
	<td> <a href="https://arxiv.org/abs/1412.6806">Springenberg et al. '15</a> </td>
	<td> ✅ </td>
	<td> <a href="https://github.com/pytorch/captum"><code>pytorch/captum</code></a> </td>
	<td>  </td>
<tr>
	<td> <a href="https://captum.ai/api/lrp.html"><b> LRP </b></a><sup><a href="#fnote2" id="fnote2-ref" data-footnote-ref="" aria-describedby="footnote-label">2</a></sup>  </td>
	<td> <a href="https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0130140">Bach et al. '15</a> </td>
	<td> ✅ </td>
	<td> <a href="https://github.com/pytorch/captum"><code>pytorch/captum</code></a> </td>
	<td>  </td>
<tr>
	<td> <a href="https://pair-code.github.io/saliency/#guided-ig"><b> Guided Integrated Gradients </b></a></td>
	<td> <a href="https://arxiv.org/abs/2106.09788">Kapishnikov et al. '21</a> </td>
	<td> </td>
	<td> <a href="https://github.com/PAIR-code/saliency"><code>PAIR-code/saliency</code></a> </td>
	<td>  </td>
<tr>
	<td> <b> Projected Gradient Descent (PGD) </b><sup><a href="#fnote3" id="fnote3-ref" data-footnote-ref="" aria-describedby="footnote-label">3</a></sup> </td>
	<td> <a href="https://arxiv.org/abs/1706.06083">Madry et al. '18</a>, <a href="https://arxiv.org/abs/2104.08782">Yin et al. '22</a> </td>
	<td> </td>
	<td> <a href="https://github.com/uclanlp/NLP-Interpretation-Faithfulness"><code>uclanlp/NLP-Interpretation-Faithfulness</code></a> </td>
	<td>  </td>
<tr>
	<td> <b> Sequential Integrated Gradients </b></td>
	<td> <a href="https://aclanthology.org/2023.findings-acl.477/">Enguehard '23</a></td>
	<td> </td>
	<td> <a href="https://github.com/josephenguehard/time_interpret"><code>josephenguehard/time_interpret</code></a> </td>
	<td> ✅ </td>
<tr>
	<td> <b> Greedy PIG </b><sup><a href="#fnote4" id="fnote4-ref" data-footnote-ref="" aria-describedby="footnote-label">4</a></sup></td>
	<td> <a href="https://openreview.net/forum?id=1GUTzm2a4v">Axiotis et al. '23</a></td>
	<td> </td>
	<td> </td>
	<td>  </td>
<tr>
	<td> <b> AttnLRP </b></td>
	<td> <a href="https://arxiv.org/abs/2402.05602">Achtibat et al. '24</a></td>
	<td> </td>
	<td> <a href="https://github.com/rachtibat/LRP-for-Transformers"><code>rachtibat/LRP-for-Transformers</code></a> </td>
	<td> ✅ </td>
<tr>
	<td> <b> Randomized Path-Integration (RPI) </b><sup><a href="#fnote5" id="fnote5-ref" data-footnote-ref="" aria-describedby="footnote-label">5</a></sup></td>
	<td> <a href="https://aclanthology.org/2024.findings-emnlp.551/">Barkan et al. '24</a></td>
	<td> </td>
	<td> <a href="https://github.com/rpiconf/rpi"><code>rpiconf/rpi</code></a> </td>
	<td> </td>
</table>

**Notes**:

1. The [Deconvolution](https://captum.ai/api/deconvolution.html) method can also be added, but it seems to perform the same procedure as Guided Backprop, so it wasn't included to avoid deduplication.

<section data-footnotes="" class="footnotes">
	<h2 id="footnote-label" class="sr-only"></h2>
	<ol dir="auto">
		<li id="fnote1"> The method was already present in inseq but was removed due to instability in the single example vs. batched setting, reintroducing it will need this problem to be fixed.</li>
		<li id="fnote2"> Custom rules for the supported architectures need to be defined in order to adapt the LRP attribution method to our use-case. An existing implementation of LRP rules for Transformer models in Tensorflow is available here: [lena-voita/the-story-of-heads](https://github.com/lena-voita/the-story-of-heads).</li>
		<li id="fnote3"> The method leverage gradient information to perform adversarial replacement, so its collocation in the gradient-based family should be reviewed.</li>
		<li id="fnote4"> Similar to Sequential Integrated Gradient, but instead of focusing on one word at a time, at every iteration the top features identified by attribution are fixed (i.e. baseline is set to identity) and the remaining ones are attributed again in the next round. </li>
		<li id="fnote5"> Integrates gradients over perturbed attention distributions </li>
	</ol>
</section>

**Comment by gsarti** [comment]
Hi @saxenarohit, in principle the Captum LRP implementation should be directly compatible with Inseq. However, the implementation is very model specific with some notable (and to my knowledge, presently unsolved) issues with skip connections, which are the bread and butter of most transformer architectures used in Inseq (see pytorch/captum#546).

I think in general to proceed with an integration we should make sure that:

1. at least the majority of Inseq-supported models would be compatible with propagation rules currently supported by Captum; and
2. the misuse of LRP for unsupported architectures would raise informative errors to prevent involuntary misuse
